File: app/src/main/java/doit/study/droid/di/AppModule.java
package doit.study.droid.di;

import android.app.Application;
import android.arch.persistence.room.Room;
import android.content.Context;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import doit.study.droid.data.source.QuestionsRepository;
import doit.study.droid.data.source.local.QuizDatabase;
import doit.study.droid.data.source.remote.QuizWebService;

@Module
public class AppModule {

    private Context context;

    public AppModule(Application app) {
        context = app;
    }

    @Provides
    @Singleton
    public Context provideAppContext() {
        return context;
    }

    @Singleton
    @Provides
    QuizDatabase provideDatabase(Context context) {
        return Room.databaseBuilder(
                context.getApplicationContext(),
                QuizDatabase.class,
                "quizDatabase.db")
                .build();
    }
    @Singleton
    @Provides
    QuestionsRepository provideRepository(QuizWebService quizWebService, QuizDatabase quizDatabase){
        return new QuestionsRepository(quizWebService, quizDatabase);
    }
}

File: app/src/main/java/doit/study/droid/di/AppModule.kt
package doit.study.droid.di

import android.app.Application
import androidx.room.Room
import com.google.android.gms.analytics.GoogleAnalytics
import com.google.android.gms.analytics.Tracker
import dagger.Module
import dagger.Provides
import doit.study.droid.R
import doit.study.droid.data.local.QuizDatabase
import javax.inject.Singleton

@Module
object AppModule {
    @Provides
    @Singleton
    fun provideAnalyticsTracker(appContext: Application): Tracker {
        val ga = GoogleAnalytics.getInstance(appContext)
        ga.enableAutoActivityReports(appContext)
        return ga.newTracker(R.xml.track_app)
    }

    @Singleton
    @Provides
    internal fun provideDatabase(appContext: Application): QuizDatabase {
        return Room.databaseBuilder(
                appContext.applicationContext,
                QuizDatabase::class.java,
                "quizDatabase.db")
                .fallbackToDestructiveMigration()
                .build()
    }
}